wfe2/wfe.go

```diff
@@ -2275,6 +2275,11 @@ func (wfe *WebFrontEndImpl) FinalizeOrder(ctx context.Context, logEvent *web.Req
 		wfe.sendError(response, logEvent, web.ProblemDetailsForError(err, "Error finalizing order"), err)
 		return
 	}
+	if updatedOrder == nil || (order.Id == 0 && order.Created == 0) && (order.RegistrationID != 0 || order.Expires != 0 || len(order.Names) != 0) {
```

Comment:
I don't think this is a correct re-implementation of `orderValid`. For example, this statement requires that both `order.Id` and `order.Created` be `0` in order to fall into the error case, whereas `orderValid` says that the order is invalid if either of those is `0`. I think you want the much simpler:
```suggestion
	if updatedOrder == nil || order.Id == 0 || order.Created == 0 || order.RegistrationID == 0 || order.Expires == 0 || len(order.Names) == 0 {
```

Reply:
Ah, okay I actually see where I got confused with this, too. Thanks.